frontend/src/components/resourceMap/graph/graphGrouping.test.ts

```diff
@@ -0,0 +1,83 @@
+import { KubeMetadata } from '../../../lib/k8s/KubeMetadata';
+import { KubeObject } from '../../../lib/k8s/KubeObject';
+import { groupGraph } from './graphGrouping';
+import { GraphEdge, KubeObjectNode } from './graphModel';
+
+describe('groupGraph', () => {
+  const nodes: KubeObjectNode[] = [
+    {
+      id: '1',
+      type: 'kubeObject',
+      data: {
+        resource: {
+          kind: 'Pod',
+          metadata: {
+            namespace: 'ns1',
+            name: 'pod1',
+            labels: { 'app.kubernetes.io/instance': 'instance1' },
+          } as any as KubeMetadata,
+        } as KubeObject,
+      },
+    },
+    {
+      id: '2',
+      type: 'kubeObject',
+      data: {
+        resource: {
+          kind: 'Pod',
+          metadata: { namespace: 'ns2', name: 'pod2' } as KubeMetadata,
+          spec: { nodeName: 'node1' },
+        } as any as KubeObject,
+      },
+    },
+    {
+      id: '3',
+      type: 'kubeObject',
+      data: {
+        resource: {
+          kind: 'Pod',
+          metadata: {
+            namespace: 'ns1',
+            name: 'pod3',
+            labels: { 'app.kubernetes.io/instance': 'instance1' },
+          } as any as KubeMetadata,
+        } as KubeObject,
+      },
+    },
+    {
+      id: '4',
+      type: 'kubeObject',
+      data: {
+        resource: {
+          kind: 'Pod',
+          metadata: { namespace: 'ns2', name: 'pod4' } as KubeMetadata,
+        } as KubeObject,
+      },
+    },
+  ];
+
+  const edges: GraphEdge[] = [{ id: 'e2', source: '2', target: '4', type: 'kubeRelation' }];
+
+  it('groups nodes by namespace', () => {
+    const groupedGraph = groupGraph(nodes, edges, { groupBy: 'namespace' });
+    const namespaces = groupedGraph.data.nodes.map(node => node.id);
+    expect(namespaces).toEqual(['Namespace-ns1', 'Namespace-ns2']);
+  });
+
+  it('groups nodes by node', () => {
+    const groupedGraph = groupGraph(nodes, edges, { groupBy: 'node' });
+    const nodeNames = groupedGraph.data.nodes.map(node => node.id);
+    expect(nodeNames).toEqual(['Node-node1', '1', '3']);
```

Comment:
Can you please explain in this test why there are three nodes in the result, but four nodes in the test data? I would expect four nodes returned.

Reply:
added comments here